src/NFT/AgentCommunication.sol

```diff
@@ -46,39 +97,43 @@ contract AgentCommunication is Ownable {
         return (amountForTreasury, amountForAgent);
     }
 
+    // We don't add `onlyRegisteredAgent` modifier here, because anyone should be able to send a message to an agent
     function sendMessage(address agentAddress, bytes memory message) public payable mustPayMoreThanMinimum {
-        // split message value between treasury and agent
+        if (!isRegisteredAgent(agentAddress)) {
+            revert IAgentRegistry.AgentNotRegistered();
+        }
+
+        // Split message value between treasury and agent
         (uint256 amountForTreasury, uint256 amountForAgent) = _calculateAmounts(msg.value);
 
         // Transfer the amounts
         (bool sentTreasury,) = treasury.call{value: amountForTreasury}("");
-        require(sentTreasury, "Failed to send Ether");
+        require(sentTreasury, "Failed to send Ether to treasury");
         (bool sentAgent,) = payable(agentAddress).call{value: amountForAgent}("");
-        require(sentAgent, "Failed to send Ether");
+        require(sentAgent, "Failed to send Ether to agent");
 
         DoubleEndedStructQueue.MessageContainer memory messageContainer =
             DoubleEndedStructQueue.MessageContainer(msg.sender, agentAddress, message, msg.value);
         DoubleEndedStructQueue.pushBack(queues[agentAddress], messageContainer);
         emit LogMessage(msg.sender, agentAddress, messageContainer.message, msg.value);
     }
 
-    function getAtIndex(address agentAddress, uint256 idx)
+    function getAtIndex(uint256 idx)
         public
         view
+        onlyRegisteredAgent
         returns (DoubleEndedStructQueue.MessageContainer memory)
     {
-        return DoubleEndedStructQueue.at(queues[agentAddress], idx);
+        return DoubleEndedStructQueue.at(queues[msg.sender], idx);
     }
```

Comment:
I would say reading should be allowed, as it's not modifying state. Messares are public.
Popping (on the other hand) should be guarded (as you implemented).

Reply:
Ah I forgot about this! Found it out while fixing errors in PMA :-D Doing here: https://github.com/gnosis/labs-contracts/pull/25